### PR TITLE
Smart contract integration

### PR DESCRIPTION
**Pull request by bhavyagor12**
Added contract connections to registerJob, companyRegister, referrerRegister and more

### CODE REVIEW

File: components/CompanyRegister.tsx [hunk 1]
```diff
@@ -23,14 +26,49 @@ const CompanyRegister = () => {
...
+        companyName,
+        companyDescription,
+        companySite,
+      ];
+      await writeAsync();
+      if (!isSuccess) {
+        toast.error("Company Register failed");
+      }
+      if (isSuccess && !isLoading) {
+        const company = await registerCompany(companyData);
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx:53]: If polybase base write fails - then we have a floating company in the SC.

These should be transactional...

For momentum happy for a ticket to be created to come back and fix this.

File: components/CompanyRegister.tsx [hunk 2]
```diff
@@ -23,14 +26,49 @@ const CompanyRegister = () => {
...
+        if (company.id) {
+          toast.success("Company Register Successful");
+        }
+      }
+    } catch (e) {
+      toast.error("Company Register failed");
```
Comment (codingnuclei) [inline components/CompanyRegister.tsx:59]:
```suggestion
      toast.error("Company Registration failed");
```

File: components/JobComponents/AddJob.tsx [hunk 3]
```diff
@@ -1,9 +1,71 @@
...
+    functionName: "registerJob",
+    args: [jobInfo.bounty],
+  });
+
+  const registerJob = async () => {
+    try {
+      const jobId = await writeAsync();
+      console.log(jobId);
+      if (isSuccess) {
+        toast.success("Register Job Successful");
```
Comment (codingnuclei) [inline components/JobComponents/AddJob.tsx:62]:
```suggestion
        toast.success("Job Registered Successfully");
```

File: components/JobComponents/AddJob.tsx [hunk 4]
```diff
@@ -1,9 +1,71 @@
...
+      }
+    } catch (e) {
+      toast.error("Register job failed");
```
Comment (codingnuclei) [inline components/JobComponents/AddJob.tsx:65]:
```suggestion
      toast.error("Job Registration failed");
```

File: components/ReferrerRegister.tsx [hunk 5]
```diff
@@ -1,33 +1,65 @@
...
+import { useAccount, useContractWrite } from "wagmi";
+import * as eth from "@polybase/eth";
+import usePolybase from "../hooks/usePolybase";
+import { recruitmentABI, recruitmentAddress } from "../src/generated";
+import { toast } from "react-toastify";
+import { toBytes, keccak256 } from "viem";
 
 const ReferrerRegister = () => {
   const [loading, setLoading] = useState(false);
+  const { db, registerReferrer } = usePolybase();
```
Comment (codingnuclei) [inline components/ReferrerRegister.tsx:12]: similar comments to CompanyRegister.tsx

File: components/ReferrerRegister.tsx [hunk 6]
```diff
@@ -1,33 +1,65 @@
...
+    functionName: "registerReferrer",
+    args: [hashEmail],
+  });
+
+  const registerReferrerSc = async () => {
+    try {
+      const referrerData = [address, name, email];
+      await writeAsync();
+      const referrer = await registerReferrer(referrerData);
+      console.log(referrer.id);
```
Comment (codingnuclei) [inline components/ReferrerRegister.tsx:45]: remove log

File: components/ReferrerRegister.tsx [hunk 7]
```diff
@@ -1,33 +1,65 @@
...
+      if (isSuccess && referrer.id) {
+        toast.success("Referrer Register Successful");
```
Comment (codingnuclei) [inline components/ReferrerRegister.tsx:47]: see previous comments

File: components/ReferrerRegister.tsx [hunk 8]
```diff
@@ -1,33 +1,65 @@
...
+      }
+    } catch (e) {
+      toast.error("Referrer Register failed");
```
Comment (codingnuclei) [inline components/ReferrerRegister.tsx:50]: see previous comments

File: hooks/usePolybase.tsx [hunk 9]
```diff
@@ -42,8 +42,9 @@ const usePolybase = () => {
     return data;
   };
 
-  const registerCompany = async (referrerData: any) => {
-    const recordData = await companiesReference.create(referrerData);
+  const registerCompany = async (companyData: any) => {
+    const recordData = await companiesReference.create(companyData);
+    console.log(recordData);
```
Comment (codingnuclei) [inline hooks/usePolybase.tsx:47]: remove logs

File: hooks/usePolybase.tsx [hunk 10]
```diff
@@ -114,6 +115,7 @@ const usePolybase = () => {
   };
 
   return {
+    db,
```
Comment (codingnuclei) [inline hooks/usePolybase.tsx:118]: see previous comments

File: helpers/index.ts [hunk 11]
```diff
@@ -9,6 +9,18 @@ export const truncateDescription = (text: string, maxWords: number): string => {
     return text;
   };
 
+
+ export const getDate = () => {
```
Comment (codingnuclei) [inline helpers/index.ts:13]: this still needed?

**Review by codingnuclei — CHANGES_REQUESTED** [review]
:fire::fire::fire:

Neat work - just a few small comments

File: helpers/index.ts (hunk 11 above)
Comment (bhavyagor12) [reply helpers/index.ts:13]: getDate needed for a job posting for polybase

File: components/CompanyRegister.tsx (hunk 1 above)
Comment (bhavyagor12) [reply components/CompanyRegister.tsx:53]: sure! pls add the ticket

Comment (codingnuclei) [reply components/CompanyRegister.tsx:53]: Tracking ticket added:

https://www.notion.so/Data-syncing-between-polybase-and-SC-needs-to-be-transactional-90237a130f134b53bda65623e5937cbf?pvs=4

File: hooks/usePolybase.tsx [hunk 12]
```diff
@@ -1,13 +1,13 @@
-import { Polybase } from "@polybase/client";
+import { Polybase, Signer } from "@polybase/client";
 import { Address } from "wagmi";
-const usePolybase = () => {
+const usePolybase = (signer:Signer) => {
```
Comment (codingnuclei) [inline hooks/usePolybase.tsx:3]:
```suggestion
const usePolybase = (signer?: Signer) => {
```
This way you don't need to pass a signer if you do not want to... so you can revert changes from earlier which added a signer which didnt have before

File: components/ReferrerRegister.tsx [hunk 13]
```diff
@@ -1,33 +1,69 @@
...
+    async (data: string) => {
+       const sig = await eth.sign(data, address);
+       return { h: "eth-personal-sign", sig };
+     })
+   ;;
   const [name, setName] = useState("");
-  const [email, setEmail] = useState("");
-
+  const [email, setEmail] = useState<string>("");
+  const [hashEmail, setHashEmail] = useState<string>("");
```
Comment (codingnuclei) [inline components/ReferrerRegister.tsx:20]:
```suggestion
  const [hashEmail, setHashEmail] = useState<`0x${string}`>();
```

File: components/ReferrerRegister.tsx [hunk 14]
```diff
@@ -1,33 +1,69 @@
...
 
   const handleNameChange = (event: any) => {
     setName(event.target.value);
   };
 
   const handleEmailChange = (event: any) => {
+    setHashEmail(String(keccak256(toBytes(event.target.value))));
```
Comment (codingnuclei) [inline components/ReferrerRegister.tsx:27]:
```suggestion
    setHashEmail(keccak256(toBytes(event.target.value)));
```

File: components/ReferrerRegister.tsx [hunk 15]
```diff
@@ -1,33 +1,69 @@
...
     setEmail(event.target.value);
   };
 
+ 
+
+  const { data, isLoading, isSuccess, writeAsync } = useContractWrite({
+    abi: recruitmentABI,
+    address: recruitmentAddress,
+    functionName: "registerReferrer",
+    args: [hashEmail as `0x${string}`],
```
Comment (codingnuclei) [inline components/ReferrerRegister.tsx:37]:
```suggestion
```

File: components/ReferrerRegister.tsx [hunk 16]
```diff
@@ -1,33 +1,69 @@
...
+  });
+
+  const registerReferrerSc = async () => {
+    try {
+      const referrerData = [address, name, email];
+      await writeAsync();
```
Comment (codingnuclei) [inline components/ReferrerRegister.tsx:43]:
```suggestion
       if(hashEmail) {
        await writeAsync({
          args: [hashEmail]
        });
      } else {
        toast.error("No email supplied");
        return;
      }
```

**Review by codingnuclei — CHANGES_REQUESTED** [review]

**Pull request merged**